charmcraft/models/charmcraft.py

```diff
@@ -0,0 +1,365 @@
+    # to load the config itself (so we're really do the validation there), and we want to support
+    # the case of a missing directory (and still load a default config structure)
+    dirpath: pathlib.Path
+    config_provided: bool = False
+
+    # this timestamp will be used in several places, even sent to Charmhub: needs to be UTC
+    started_at: datetime.datetime
+
+
+class Ignore(ModelConfigDefaults):
+    """Definition of `analysis.ignore` configuration."""
+
+    attributes: List[AttributeName] = []
+    linters: List[LinterName] = []
+
+
+class AnalysisConfig(ModelConfigDefaults, allow_population_by_field_name=True):
+    """Definition of `analysis` configuration."""
+
+    ignore: Ignore = Ignore()
+
+
+class Links(ModelConfigDefaults):
+    """Definition of `links` in metadata."""
+
+    contact: Optional[Union[pydantic.StrictStr, List[pydantic.StrictStr]]]
+    documentation: Optional[pydantic.AnyHttpUrl]
+    issues: Optional[Union[pydantic.AnyHttpUrl, List[pydantic.AnyHttpUrl]]]
+    source: Optional[Union[pydantic.AnyHttpUrl, List[pydantic.AnyHttpUrl]]]
+    website: Optional[Union[pydantic.AnyHttpUrl, List[pydantic.AnyHttpUrl]]]
+
+
+class CharmcraftConfig(
+    ModelConfigDefaults,
+    validate_all=False,
+    alias_generator=lambda s: s.replace("_", "-"),
+):
+    """Definition of charmcraft.yaml configuration."""
+
+    # this needs to go before 'parts', as it used by the validator
+    project: Project
+
+    metadata_legacy: bool = False
+
+    type: str
+    name: Optional[pydantic.StrictStr]
+    summary: Optional[pydantic.StrictStr]
+    description: Optional[pydantic.StrictStr]
+    charmhub: CharmhubConfig = CharmhubConfig()
+    parts: Optional[Dict[str, Any]]
+    bases: Optional[List[BasesConfiguration]]
+    analysis: AnalysisConfig = AnalysisConfig()
+    actions: Optional[JujuActions]
+    assumes: Optional[List[Union[str, Dict[str, List[str]]]]]
+    containers: Optional[Dict[str, Any]]
+    devices: Optional[Dict[str, Any]]
+    title: Optional[pydantic.StrictStr]
+    extra_bindings: Optional[Dict[str, Any]]
+    peers: Optional[Dict[str, Any]]
+    provides: Optional[Dict[str, Any]]
+    requires: Optional[Dict[str, Any]]
+    resources: Optional[Dict[str, Any]]
+    storage: Optional[Dict[str, Any]]
+    subordinate: Optional[bool]
+    terms: Optional[List[str]]
+    links: Optional[Links]
+    config: Optional[JujuConfig]
+
+    @pydantic.validator("type")
+    def validate_charm_type(cls, charm_type):
+        """Verify charm type is valid with exception when instantiated without YAML."""
+        if charm_type not in ["bundle", "charm"]:
+            raise ValueError("must be either 'charm' or 'bundle'")
+        return charm_type
+
+    @pydantic.validator("name", pre=True, always=True)
+    def validate_name(cls, name, values):
+        """Verify charm name is valid with exception when instantiated without YAML."""
+        if values.get("type") == "charm" and not name:
```

Comment:
Given the number of these, could we make separate models for if it's a charm or a bundle and a factory for generating the correct model?

Reply:
That's possible, just another refactor.